designs/node-overlay-dra.md

```diff
@@ -0,0 +1,207 @@
+# NodeOverlay Integration for Dynamic Resource Allocation
+
+## TODO
+
+- [ ] Concrete schema example
+- [ ] NodeOverlay conflict resolution definition
+- [ ] Integration with NodeClaim initialization check
+
+## Background
+
+When Karpenter constructs a NodeClaim for a given set of pods, it uses information from the CloudProvider to determine
+the resources that will exist on a Node if one is created for a given instance type. That is then used to determine
+which instance types will satisfy a given set of requests. CloudProviders can get this information from any number of
+sources, for example EC2's `DescribeInstanceTypes` API, and they use their knowledge of well-known device plugins to
+determine the resulting Node's allocatable resources. The drawback of this approach is that each CloudProvider needs to
+build integrations for specific drivers, and users are limited to those integrations. `NodeOverlays` were introduced to
+close that gap by allowing end users to explicitly inform Karpenter of the resources that will be made available for a
+given instance type. For example, the following `NodeOverlay` will add the `smarter-device/fuse` device to `c5.large`
+instances from the NodePool `fuse-np`.
+
+```yaml
+kind: NodeOverlay
+apiVersion: karpenter.sh/v1alpha1
+spec:
+  requirements:
+  - key: karpenter.sh/nodepool
+    operator: In
+    values: ['fuse-np']
+  - key: node.kubernetes.io/instance-type
+    operator: In
+    values: ['c5.large']
+  capacity:
+    smarter-device/fuse: 1
+```
+
+We face the same challenge with DRA: CloudProviders will need to be made aware of the `ResourceSlices` that will be
+published when a given instance type registers with the cluster as a node. This RFC proposes extending the existing
+NodeOverlay API to support specifying `ResourceSlices`  in addition to conventional extended resources.
+
+## Non-goals
+- Support for provisioning devices which aren't associated directly with a node
+	- As a node autoscaler, I would consider this out-of-scope for Karpenter initial support. The integration in Karpenter's
+      scheduler should be able to support any `ResourceSlice` which has already been provisioned, but that's a separate
+      matter from supporting the provisioning of new devices.
+
+## General API Shape
+
+At a high level, we will want to add a new field to `NodeOverlay`, `resourceSlices`, which will be a list of
+`ResourceSlices` to associate with any matching instance types. This wouldn't be a breaking change to the API, so it
+would remain `v1alpha1`.
+
+
+There is some nuance to they type of the `resourceSlices` field. There are three options:
+- A list of unstructured objects
+- The upstream v1 schema
+- A custom `ResourceSlice` schema
```

Comment:
@towca what were some of the considerations that went into how CAS represents ResourceSlice data in its simulator flow?

Reply:
IIUC CAS uses the upstream scheduler implementation directly for it's simulation, so I assume it uses the upstream schema. If we used the upstream scheduler, that's what we'd do as well.